File: core/src/com/xenoage/zong/core/music/VoiceElement.java
package com.xenoage.zong.core.music;

import com.xenoage.utils.math.Fraction;
import com.xenoage.zong.core.position.MPElement;


/**
 * Interface for all classes that are child of a voice
 * and have a duration.
 * 
 * These are chords and rests.
 * 
 * @author Uli Teschemacher
 * @author Andreas Wenger
 */
public interface VoiceElement
	extends MPElement
{
	
	/**
   * Gets the duration of this element.
   */
  public Fraction getDuration();
  
  
  /**
   * Sets the duration of this element.
   */
  public void setDuration(Fraction duration);
  
  
  /**
	 * Gets the parent voice, or null.
	 */
	@Override public Voice getParent();
	
	
	/**
	 * Sets the parent voice, or null.
	 */
	public void setParent(Voice parent);
	
}

File: core/src/com/xenoage/zong/core/music/VoiceElement.kt
package com.xenoage.zong.core.music

import com.xenoage.zong.core.music.util.Duration
import com.xenoage.zong.core.position.MPContainer
import com.xenoage.zong.core.position.MPElement


/**
 * Interface for all classes that are child of a voice
 * and have a duration.
 *
 * These are chords and rests.
 */
interface VoiceElement : MPElement, MPContainer {

	/** The duration of this element. */
	var duration: Duration

	/** Back reference: The parent voice, or null if not part of a voice. */
	override var parent: Voice?

	/** All children have the same musical position as this element. */
	override fun getChildMP(child: MPElement) = mp

}
